tune_sklearn/tune_basesearch.py

```diff
@@ -479,29 +539,54 @@ def _fit(self, X, y=None, groups=None, **fit_params):
+        # In single metric evaluation, refit_metric is "score"
+        if self.refit or not self.is_multi:
+            # If callable, refit is expected to return the index of the best
+            # parameter set.
+            if callable(self.refit):
+                self.best_index = self.refit(self.cv_results_)
+                if not isinstance(self.best_index, numbers.Integral):
+                    raise TypeError("best_index returned is not an integer")
+                if (self.best_index < 0
+                        or self.best_index >= len(self.cv_results_["params"])):
+                    raise IndexError("best_index index out of range")
+            else:
+                self.best_index = self.cv_results_["rank_test_%s" %
+                                                   base_metric].argmin()
+                self.best_score = self.cv_results_[
+                    "mean_test_%s" % base_metric][self.best_index]
             best_config = analysis.get_best_config(
                 metric=metric, mode="max", scope="last")
             self.best_params = self._clean_config_dict(best_config)
 
-            self.best_estimator_ = clone(self.estimator)
+        if self.refit:
+            base_estimator = clone(self.estimator)
             if self.early_stop_type == EarlyStopping.WARM_START_ENSEMBLE:
                 logger.info("tune-sklearn uses `n_estimators` to warm "
                             "start, so this parameter can't be "
                             "set when warm start early stopping. "
                             "`n_estimators` defaults to `max_iters`.")
-                if check_is_pipeline(self.estimator):
-                    cloned_base_estimator = self.best_estimator_.steps[-1][1]
-                    cloned_base_estimator.set_params(
+                if check_is_pipeline(base_estimator):
+                    cloned_final_estimator = base_estimator.steps[-1][1]
+                    cloned_final_estimator.set_params(
                         **{"n_estimators": self.max_iters})
                 else:
                     self.best_params["n_estimators"] = self.max_iters
-            self.best_estimator_.set_params(**self.best_params)
-            self.best_estimator_.fit(X, y, **fit_params)
-
-            best_result = analysis.get_best_trial(
-                metric=metric, mode="max", scope="last").last_result
-            self.best_score = float(best_result[metric])
+            # we clone again after setting params in case some
+            # of the params are estimators as well.
+            self.best_estimator = clone(
+                base_estimator.set_params(**self.best_params))
+            refit_start_time = time.time()
+            if y is not None:
+                self.best_estimator.fit(X, y, **fit_params)
+            else:
+                self.best_estimator.fit(X, **fit_params)
+            refit_end_time = time.time()
+            self.refit_time = refit_end_time - refit_start_time
```

Comment:
Can you be sure to set an instance variable in the constructor for this?

Reply:
Will do!